### PR TITLE
Clean up MIOSR

### PR DESCRIPTION
**Pull request by wesg52**
- Improve docstring
- Delete convenience code for experiments
- Add MIOSR to equality constraint tests

### CODE REVIEW

**Comment by Jacob-Stevens-Haas** [comment]
Hey @wesg52 , just want you to know we haven't forgotten about this.  It's near the top of my priority list, and I think I should get to it by the weekend, if not later this week.  Thanks for your patience.

**Comment by Jacob-Stevens-Haas** [comment]
Firstly, you write wonderfully.  I really enjoyed reading the paper.  It's exciting to see the opportunities that MIO provides for SINDy.

Secondly, can you close this PR and open one from `wesg52:miosr` directly into `dynamicslab:master`, rather than back into `dynamicslab:miosr`?  I'm reading through the code now & preparing questions/comments but it's easier to add comments and review writ large that way.  Sorry.

**Comment by wesg52** [comment]
First, thank you! I am glad you enjoyed and am excited to see people more knowledgable about dynamics find good use cases.

Second, I changed the target branch to `dynamicslab:master`, I assume that is sufficient.

Third, it seems like the free version of gurobipy worked fined in the CI test suite so that should hopefully not be an issue.

File: pysindy/optimizers/miosr.py [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+        return beta.X
+
+    def _reduce(self, x, y):
+        """
+        Runs MIOSR either per dimension or jointly on all dimensions.
+
+        Assumes an initial guess for coefficients and support are saved in
+        ``self.coef_`` and ``self.ind_``.
+        """
+        regress_jointly = self.target_sparsity is not None
```
Comment (Jacob-Stevens-Haas) [inline pysindy/optimizers/miosr.py:217]: Nit: These kind of single-use boolean flags have a tendency to migrate further away from where they matter as later changes add lines of code.  Just inline it as `if target_sparsity is not None:` and add `# regress_jointly` as a comment.  (if that is, in fact, the desired behavior... see comment on the class docstring).

**Review by Jacob-Stevens-Haas — CHANGES_REQUESTED** [review]
Nicely done - this seems to work and fit in pysindy well.  I added a few style requests, and it would help if you added a test of the constrained version (like `test_trapping_inequality_constraints`).  Only real confusion is over how `target sparsity` and `group sparsity` interact - I'm not sure they're behaving as described w.r.t `initial_guess`, adding constraints, and whether targets are regressed jointly or independently.

RE: earlier comments you made in #215:
* "Added a default sparsity which fixed the default argument tests (though not clear if that is actually the 'right' thing to do)."  That's a good point - there's no innate reason why optimizers *need* a zero-argument form, but there's also no guarantee that any optimizer fits well with default sparsity arguments. 
* Don't worry about `test_complexity_parameter[SR3-threshold]` - it uses random initial data and occasionally fails.
* When you said `test_complexity` fails, did you mean adding `MIOSR` to `test_optimizers_complexity.test_complexity`?  If so - it just tests that when optimizers promote more sparsity, they get it.  No action needed.
* "how best to enable users to have fine grained control over the gurobi model..."  Do we really need to allow users fine grained control over the gurobi model?  If you think so, you could add a `gurobi_kwargs` argument, but this seems like a case of "No is temporary, but yes is forever"

File: pysindy/optimizers/miosr.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+            coefs[non_active_ixs] = 0
+            self.coef_ = coefs.reshape(r, d)
+            self.ind_ = (np.abs(self.coef_) > 1e-6).astype(int)
+        else:
+            for i in range(r):
+                k = self.group_sparsity[i]
+                warm_start = (
+                    None if self.initial_guess is None else self.initial_guess[[i], :]
+                )
+                coef_i = self._regress(x, y[:, [i]], k, warm_start=warm_start)
```
[the remark this answers is not in the document]
Comment (wesg52) [reply pysindy/optimizers/miosr.py:238]: No, good catch.

File: pysindy/optimizers/miosr.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+
+        Bertsimas, D. and Gurnee, W., 2022. Learning Sparse Nonlinear Dynamics
+        via Mixed-Integer Optimization. arXiv preprint arXiv:2206.00176.
+
+    Parameters
+    ----------
+    target_sparsity : int, optional (default None)
+        The maximum number of nonzero coefficients across all dimensions.
+        If set, and `group_sparsity` is not set, the model will fit all
+        dimensions jointly, potentially reducing statistical efficiency.
```
[the remark this answers is not in the document]
Comment (wesg52) [reply pysindy/optimizers/miosr.py:34]: Yes docstring is wrong, good catch!

**Comment by wesg52** [comment]
Sorry for the delay. Thank you for the detailed look.

`group_sparsity` and `target_sparsity` are allowed together. However, I would expect all use cases where both are defined to have the property that `target_sparsity < sum(group_sparsity)` otherwise the `target_sparsity` constraint is redundant.

In general, `target_sparsity || constraints => regress_jointly` (technically there are some conditions where you could have per dimension constraints but I don't expect this to be an important use case.)

I believe I implemented all your suggestions and all tests are passing.

Regarding "how best to enable users to have fine grained control over the gurobi model..." I agree we don't need to implement this feature, given it is simple enough to just copy miosr and reimplement whatever modeling logic is required within `make_model()`.

File: pysindy/optimizers/__init__.py [hunk 4]
```diff
@@ -1,6 +1,7 @@
 from .base import BaseOptimizer
 from .constrained_sr3 import ConstrainedSR3
 from .frols import FROLS
+from .miosr import MIOSR
```
[the remark this answers is not in the document]
Comment (Jacob-Stevens-Haas) [reply pysindy/optimizers/__init__.py:4]: Oops I missed something.  This raises an error if people haven't installed `pysindy[miosr]` every time they import `pysindy`.  Same thing in optimizers/\_\_init\_\_.py: any import of `pysindy` triggers importing `pysindy.optimizers`, which then triggers the import of `pysindy.optimizers.miosr` and `gurobipy`.

Lazy imports might fix this in the future, but the [PEP](https://peps.python.org/pep-0690/) is still in draft.

So in the interim, guard `from .optimizers import MIOSR` in pysindy/\_\_init\_\_.py and `from .miosr import MIOSR` in pysindy/optimizers/\_\_init\_\_.py in a 

```python
try:
    <import statement>
except ImportError:
    pass
```

**Review by Jacob-Stevens-Haas — CHANGES_REQUESTED** [review]
Thanks for the fixes :)!  Only a few things.
- One oversight I made on the import mechanics 
- Update docstring default for `target_sparsity` to match `__init__`
- The leftover print statement

**Comment by wesg52** [comment]
Alright, I think that does it! Let me know if there is anything else.

**Review by Jacob-Stevens-Haas — APPROVED** [review]
LGTM!  Thanks again @wesg52

**Comment by akaptano** [comment]
Great job with this. I think the last thing that would be good to add is a simple example of the MIOSR functionality added to the "Optimizers" section of the Example 1 Jupyter notebook. There are some features of this algorithm that are unique, like the target_sparsity and group_sparsity, that would be good to demonstrate. Moreover, if we showcase it in Example 1, more people will see it!

**Review by znicolaou — COMMENTED** [review]
This looks really great to me. I am checking out the arxiv paper to learn about the miosr approach, but it all looks well-reasoned to me so far. I agree that it would be helpful to create a jupyter notebook showing off the functionality of the new optimizer, especially so new users can test it out and start using it on their data.

**Comment by wesg52** [comment]
Thanks all! 

I could imagine demonstrating
- target_sparsity vs. group_sparsity
- adding constraints
- running with verbose to print the gurobi output (and link docs explaining what it means)
Anything else you think is especially important or anything from this list you would cut?

(Regardless, I will get to this Friday or Saturday.)

**Comment by akaptano** [comment]
This is a good list. My meager understanding is that MIOSR should beat ConstrainedSR3 on some problems, and this would be nice demonstration if you can cook up a dynamical system example where it does outperform ConstrainedSR3. I suppose this might require scanning all the MIOSR and ConstrainedSR3 hyperparameters to fully conclude though? 

I think the most important things to show is that (1) we have now two algorithms in the code which can handle constraints (well three including trappingSR3 but it is derived from ConstrainedSR3) and (2) it is the only algorithm that can switch between these different notions of sparsity. 

Thanks for your work on this!!

**Comment by wesg52** [comment]
Indeed, it would beat SR3 on some problems but I think for keeping with the simplicity and speed of the example notebook, I opted to just stay consistent with the other examples. Let me know if you'd like any changes to this!

**Comment by akaptano** [comment]
Looks great. I'm basically ready to approve and merge this, but last thing I was wondering is if we could begin the MIOSR section of the Example 1 notebook with something like an if statement that checks if gurobipy is installed (and if not, skips running this section), since the notebook will stop and give an error if the user hasn't done this.

**Comment by akaptano** [comment]
Hi Wes, do you mind fixing the conflicts with 1_feature_overview? You can use the options listed here to do that. I tried doing this and can successfully put everything together, but I do not have the gurobipy license, so when I rerun the example 1 notebook it gets rid of all the nice example output you have.

**Comment by akaptano** [comment]
Alternatively, just rerun the example script on the akaptano_miosr branch I just pushed.

**Comment by Jacob-Stevens-Haas** [comment]
FYI it looks like the only conflicts are in the base64 encoded images and the conflict markers get put in stderr cells in the notebook json.  Re-running the notebook clears them.  (However, that might only work with certain diff drivers in git - I'm using nbdiff)

**Comment by wesg52** [comment]
Hi Alan, you automatically get the reduced license when you `pip install gurobipy` so I would suggest just doing that (and I imagine you will want it in the future to play with MIOSR anyway)

**Comment by akaptano** [comment]
I am finding that I have gurobipy, but it complains I do not have a license with a GurobiError. So "import gurobipy" works fine but then using it gives me a GurobiError saying I do not have a license. Not sure how to reconcile that with what you said... I think I downloaded gurobipy a long while ago though. I am using version 9.1.2. This might be just a problem on my end though. Let me know what y'all think and sorry this has been a slow PR to finish!

**Comment by wesg52** [comment]
Ah ok if you got it a while ago (e.g. when 9.1.2 came out) then your license has probably expired.

First, perhaps add a model within the try block (`m = gp.Model()`) which should fail if you don't have a valid license (as per [this](https://support.gurobi.com/hc/en-us/articles/360034784572-How-do-I-check-for-a-valid-license-file-) doc) so that the example notebook does not raise an exception in this case.

Second, update gurobi (I use somethings from the matrix interfaces in the 9.5 update so it would have failed even if you had a license) and 'renew' your license as described [here](https://support.gurobi.com/hc/en-us/articles/360038967271-How-do-I-renew-my-free-individual-academic-or-free-trial-license-).

**Comment by akaptano** [comment]
Closing this pull request and merging akaptano-miosr directly into the main branch. Thanks very much for your excellent work Wes and I'll reach out to Steve about having you present on an upcoming Friday, if you are still interested.

**Pull request closed without merging**